### PR TITLE
Values are now flattened in providers 

### PR DESCRIPTION
**Pull request by loichuder**
Work towards https://github.com/silx-kit/h5web/issues/616#issuecomment-828238108

This means that the `valuesStore` now only stores flattened values.

There are several points I would like to discuss that I highlight below.

### CODE REVIEW

File: src/h5web/providers/hsds/hsds-api.ts [hunk 1]
```diff
@@ -109,14 +109,12 @@ export class HsdsApi extends ProviderApi {
...
     // HSDS does not reduce the number of dimensions when selecting indices
-    // This is an issue when trying to slice nD arrays
-    // Therefore the dimension reduction is done "manually" here with a flat
-    const numberIndices = selection.match(/\d+/gu)?.length;
-    if (!numberIndices) {
-      return value;
+    // Therefore the flattening must be done on all dimensions regardless of the selection
+    if (hasArrayShape(entity)) {
+      return (value as unknown[]).flat(entity.shape.length - 1);
     }
```
Comment (loichuder) [inline src/h5web/providers/hsds/hsds-api.ts:115]: https://github.com/HDFGroup/hsds/issues/88 works in our favour here as `value` will always have the same number of dimensions of the original dataset, even when slicing

File: src/h5web/providers/jupyter/jupyter-api.ts [hunk 2]
```diff
@@ -161,4 +163,14 @@ export class JupyterStableApi extends ProviderApi {
 
     return contents;
   }
+
+  protected flattenValue(value: unknown, entity: Dataset, selection?: string) {
+    if (!hasArrayShape(entity)) {
+      return value;
+    }
+
+    const slicedDims = selection?.split(',').filter((s) => s.includes(':'));
```
Comment (loichuder) [inline src/h5web/providers/jupyter/jupyter-api.ts:172]: In Jupyter, the returned `value` dimensions change with the slicing. 

I detect when a dimension is conserved by checking for `:`.

File: src/h5web/providers/jupyter/jupyter-dev-api.ts [hunk 3]
```diff
@@ -32,7 +33,15 @@ export class JupyterDevApi extends JupyterStableApi {
...
-    return this.fetchData(params);
+    const { path, selection } = params;
+    const [value, entity] = await Promise.all([
+      this.fetchData(params),
+      this.getEntity(path),
+    ]);
+
+    assertDataset(entity);
+
+    return this.flattenValue(value, entity, selection);
```
Comment (loichuder) [inline src/h5web/providers/jupyter/jupyter-dev-api.ts:44]: As I need to know the dataset dimensions to flatten, I need to fetch the entity along with the data...

This was not an issue for `JupyterStable` and `Hsds` as the entity was already fetched for other reasons (complex type parsing in Jupyter and dataset assertion in Hsds)

This is not ideal but won't be a problem anymore if we move the flattening in the backend

File: src/h5web/providers/mock/mock-api.ts [hunk 4]
```diff
@@ -39,28 +43,27 @@ export class MockApi extends ProviderApi {
...
+    const mappedArray = applyMapping(
+      dataArray,
+      selection
         .split(',')
-        .map((s) => (s === ':' ? null : Number.parseInt(s, 10)))
+        .map((s) => (s === ':' ? 'x' : Number.parseInt(s, 10)))
     );
 
-    return unpack(dataView);
+    return mappedArray.data;
```
Comment (loichuder) [inline src/h5web/providers/mock/mock-api.ts:66]: I cannot use `dataView.data` as it holds the data of the full `dataArray`. I therefore make use for `applyMapping` which allow me to remove `unpack` from the project dependencies.

File: src/h5web/vis-packs/core/utils.ts [hunk 5]
```diff
@@ -305,7 +305,7 @@ export function getBaseArray<T>(
   value: T[] | undefined,
   rawDims: number[]
 ): NdArray<T> | undefined {
-  return value && ndarray<T>(value.flat(rawDims.length - 1) as T[], rawDims);
+  return value && ndarray<T>(value, rawDims);
```
Comment (loichuder) [inline src/h5web/vis-packs/core/utils.ts:308]: `value` is now expected to be flat

File: src/h5web/providers/mock/mock-api.ts [hunk 6]
```diff
@@ -39,28 +43,27 @@ export class MockApi extends ProviderApi {
       await this.cancellableDelay(params);
     }
 
-    const { value } = dataset;
+    const { value: rawValue } = dataset;
+    const value = hasArrayShape(dataset)
+      ? (rawValue as unknown[]).flat(dataset.shape.length - 1)
+      : rawValue;
```
Comment (loichuder) [inline src/h5web/providers/mock/mock-api.ts:49]: Note that we could remove this step if we store only flat arrays as mock data

File: src/h5web/providers/mock/mock-api.ts (hunk 4 above)
Comment (axelboc) [reply src/h5web/providers/mock/mock-api.ts:66]: Clever. Seems a bit hacky to replace `:` with `x`, though. Maybe `applyMapping` could accept the type `(':' | number)[]`? Don't think you'd have to change the implementation.

File: src/h5web/vis-packs/core/complex/MappedComplexVis.tsx [hunk 7]
```diff
@@ -45,13 +44,7 @@ function MappedComplexVis(props: Props) {
...
   const { visType } = useComplexConfig((state) => state, shallow);
 
-  const [slicedDims, slicedMapping] = useMemo(
-    () => [
-      dims.filter((_, i) => isAxis(dimMapping[i])),
-      dimMapping.filter((dim) => isAxis(dim)),
-    ],
-    [dims, dimMapping]
-  );
+  const [slicedDims, slicedMapping] = useSliceDimsAndMapping(dims, dimMapping);
```
Comment (axelboc) [inline src/h5web/vis-packs/core/complex/MappedComplexVis.tsx:47]: `useSlicedDimsAndMapping` (missing a `d`)

**Review by axelboc — APPROVED** [review]
Nicely done

File: src/h5web/vis-packs/core/utils.ts [hunk 8]
```diff
@@ -305,24 +305,26 @@ export function getBaseArray<T>(
...
+  mapping: (number | Axis | ':')[]
 ): NdArray<T> | undefined {
   if (!baseArray) {
     return undefined;
   }
 
   const isXBeforeY =
-    mapping.includes('y') && mapping.indexOf('x') < mapping.indexOf('y');
+    mapping.includes('y') &&
+    mapping.includes('x') &&
```
Comment (axelboc) [inline src/h5web/vis-packs/core/utils.ts:326]: Ah nice 👏

**Review by axelboc — APPROVED** [review]

**Pull request merged**